### PR TITLE
Only close eventFD in EpollEventLoop when it cannot be written to.

### LINKED ISSUE #9362: Shutdown race in EpollEventLoop
In 4.1.38, there is an fd race in EpollEventLoop.  The loop thread, while trying to shutdown, closes the eventFd.    The outside thread that calls shutdownGracefully, tries to wake up the loop to notice that it wants it to shutdown.   There is a race where the wakeup() call can write to an eventfd that has been closed already, resulting in the possibility of writing to the wrong fd.

The sequence of events loops like:

1.  T1:  eventLoop.shutdownGracefully()
2.  T1:  state = ST_SHUTDOWN
2.  T2:  reads state == ST_SHUTDOWN
2.  T1:  Enters  EpollEventLoop.wakeup()
2.  T2:  Exits loop, enters EpollEventLoop.cleanup() 
2.  T2:  Closes eventFd.
2.  T1: Wins race to by setting wakenUp to 1
2.  T1: Calls Native.eventFdWrite(eventFd.intValue(), 1L);

The easiest way I can see to fix this race is to use a synchronized block in wakeup(), after the thread has won the wakenUp = 1 race:

```java
    @Override
    protected void wakeup(boolean inEventLoop) {
        if (!inEventLoop && WAKEN_UP_UPDATER.getAndIncrement(this) == 0) {
            // write to the evfd which will then wake-up epoll_wait(...)
            synchronized (eventFd) {
                if (!isShutdown()) {
                    Native.eventFdWrite(eventFd.intValue(), 1L);
                }
            }
        }
    }
```

 Additionally, in EpollEventLoop.cleanup, acquire a lock before closing the eventFd, synchronize the close:

```java
            synchronized (eventFd) {
                try {
                    eventFd.close();
                } catch (IOException e) {
                    logger.warn("Failed to close the event fd.", e);
                }
            }
```

This is kind of unfortunate, but the race makes it possible that another FD is opened after closing the eventFd, and the wakeUp call writes to the wrong FD.  The main downside of my proposed patch is probably the risk of deadlocks (because I dont know what locks I hold), as well as some performance penalty.   

Ideas on how to not pay this cost are welcome, but I think the danger is enough to merit fixing this.

cc @ejona86 @normanmaurer

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
Motivation:
Writing to the eventFD is racy with close.   If the event loop is shutdown
and the fd closed, the fd number could be reused.   An in progress call to wakeup()
may write to the newly allocated fd, causing corruption.

Modification:
Make wakeup() indicate if a write is in progress or has completed.  The normal
loop will clear this indicator.  A shutdown now waits for the pending fd write to
finish.

Result:
No TSAN races reported.

Fixes #9362

### CODE REVIEW

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java [hunk 1]
```diff
@@ -177,9 +182,13 @@ NativeDatagramPacketArray cleanDatagramPacketArray() {
...
     @Override
     protected void wakeup(boolean inEventLoop) {
-        if (!inEventLoop && WAKEN_UP_UPDATER.getAndSet(this, 1) == 0) {
+        if (!inEventLoop && WAKEN_UP_UPDATER.compareAndSet(this, WAKEUP_ALLOWED, WAKEUP_INPROGRESS)) {
             // write to the evfd which will then wake-up epoll_wait(...)
-            Native.eventFdWrite(eventFd.intValue(), 1L);
+            try {
+                Native.eventFdWrite(eventFd.intValue(), 1L);
+            } finally {
+                WAKEN_UP_UPDATER.compareAndSet(this, WAKEUP_INPROGRESS, WAKEUP_NOT_ALLOWED);
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java]: its a bit sad that we now need another CAS :/

**Comment by normanmaurer** [comment]
@netty-bot test this please

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java [hunk 2]
```diff
@@ -73,7 +77,8 @@ public int get() throws Exception {
             return epollWaitNow();
         }
     };
-    @SuppressWarnings("unused") // AtomicIntegerFieldUpdater
+    // wakenUp is a packed field, where the bottom bit indicates if the loop is trying to exit
+    // The high bits indicate if there are multiple non-loop threads trying to wake up the loop.
```
[the remark this answers is not in the document]
Comment (carl-mastrangelo) [reply transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java]: Oops, that should not have survived.  That was from a previous version.

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java [hunk 3]
```diff
@@ -454,6 +463,15 @@ protected void cleanup() {
...
             }
             try {
+                // There are three possibilities:
+                // a.  wakenUp is WAKEUP_ALLOWED, which means no pending write
+                // b.  wakenUp is WAKEUP_NOT_ALLOWED, which means no pending writes
+                // c.  wakenUp is WAKEUP_INPROGRESS, which means will it will eventually reach WAKEUP_NOT_ALLOWED.
+                WAKEN_UP_UPDATER.compareAndSet(this, WAKEUP_ALLOWED, WAKEUP_NOT_ALLOWED);
+
+                // Busy loop until the last wakeup() is out of the critical section.
+                while (wakenUp != WAKEUP_NOT_ALLOWED) { /* spin */ }
```
[the remark this answers is not in the document]
Comment (carl-mastrangelo) [reply transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java]: Done.

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java [hunk 4]
```diff
@@ -454,6 +459,15 @@ protected void cleanup() {
                 logger.warn("Failed to close the epoll fd.", e);
             }
             try {
+                // There are three possibilities:
+                // a.  wakenUp is WAKEUP_ALLOWED, which means no pending write
+                // b.  wakenUp is WAKEUP_NOT_ALLOWED, which means no pending writes
+                // c.  wakenUp is WAKEUP_INPROGRESS, which means will it will eventually reach WAKEUP_NOT_ALLOWED.
+                if (!WAKEN_UP_UPDATER.compareAndSet(this, WAKEUP_ALLOWED, WAKEUP_NOT_ALLOWED)) {
+                    // Busy loop until the last wakeup() is out of the critical section.
+                    while (wakenUp != WAKEUP_NOT_ALLOWED) { /* spin */ }
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java]: Maybe not a big deal but should we add a sleep(50) or so to reduce CPU costs ?

**Comment by njhill** [comment]
@carl-mastrangelo I'm actually not sure this is watertight yet, the below is  maybe unlikely but I think possible... WDYT?

- EL is awake, `wakenUp == WA`
- T1 submits task and CAS succeeds, now `wakenUp == WIP`, proceeds to call eventfd write
- EL completes task processing, sets `wakenUp = WA` prior to going to sleep
- T2 submits task and CAS succeeds, now `wakenUp == WIP`
- T1 finishes eventfd write, successfully CASes `wakenUp` from `WIP` to `WNA`
- Channel was closed at some point, EL proceeds to close eventfd, not blocked because state is `WNA`
- fd is reassigned, T2 calls eventfd write

**Comment by njhill** [comment]
Here is a possible alternative that avoids the race described above: b397f45cd4d39ee05e26eeba1d0f411ac015ac77, and does not use any CAS for wakeup

I don't think it's quite what @ejona86 was suggesting in #9362, but might be similar. And it's not clear to me how polling for EAGAIN or a non-EAGAIN would avoid the race - I don't think either tells you for sure that there is or isn't some outstanding write in progress?

It does mean typically reading from the fd during close, but I'm not sure that's a big deal. Closing shouldn't be high frequency or on critical path (unlike wakeup operation for example), and it already involves some syscalls to close all the fds.

**Comment by normanmaurer** [comment]
@ejona86 please check https://github.com/netty/netty/commit/36ff672b808fd5e17fb908fd22a2ce4f9cae97eb.

**Comment by normanmaurer** [comment]
This was fixed by https://github.com/netty/netty/pull/9535

**Pull request closed without merging**